src/app/clusters/on-off-server/on-off-server.h

```diff
@@ -17,19 +17,83 @@
- * normal channels or as a result of a level change.
+/**********************************************************
+ * Defines and Macros
+ *********************************************************/
+
+static constexpr uint8_t UPDATE_TIME_MS      = 100;
+static constexpr uint16_t TRANSITION_TIME_1S = 10;
+
+static constexpr uint16_t MAX_TIME_VALUE = 0xFFFF;
+static constexpr uint8_t MIN_TIME_VALUE  = 1;
+
+/**
+ * @brief
  *
- * @param endpoint   Ver.: always
- * @param command   Ver.: always
- * @param initiatedByLevelChange   Ver.: always
  */
-EmberAfStatus emberAfOnOffClusterSetValueCallback(chip::EndpointId endpoint, uint8_t command, bool initiatedByLevelChange);
+class OnOffServer
+{
+public:
+    /**********************************************************
+     * Functions Definitions
+     *********************************************************/
+
+    static OnOffServer & Instance();
+
+    bool offCommand();
+    bool onCommand();
+    bool toggleCommand();
+    void initOnOffServer(chip::EndpointId endpoint);
+    bool offWithEffectCommand(uint8_t effectId, uint8_t effectVariant);
+    bool OnWithRecallGlobalSceneCommand();
+    bool OnWithTimedOffCommand(chip::BitFlags<chip::app::Clusters::OnOff::OnOffControl> onOffControl, uint16_t onTime,
+                               uint16_t offWaitTime);
+    void updateOnOffTimeCommand(chip::EndpointId endpoint);
+    EmberAfStatus setOnOffValue(chip::EndpointId endpoint, uint8_t command, bool initiatedByLevelChange);
+
+private:
+    /**********************************************************
+     * Functions Definitions
+     *********************************************************/
+
+#ifdef ZCL_USING_ON_OFF_CLUSTER_START_UP_ON_OFF_ATTRIBUTE
+    bool areStartUpOnOffServerAttributesTokenized(chip::EndpointId endpoint);
+#endif // ZCL_USING_ON_OFF_CLUSTER_START_UP_ON_OFF_ATTRIBUTE
+    EmberEventControl * getEventControl(chip::EndpointId endpoint);
+    EmberEventControl * configureEventControl(chip::EndpointId endpoint);
+
+    /**********************************************************
+     * Attributes Decleration
+     *********************************************************/
+
+    static OnOffServer instance;
+    EmberEventControl eventControls[EMBER_AF_ON_OFF_CLUSTER_SERVER_ENDPOINT_COUNT];
+};
+
+struct OnOffEffect
+{
+    using OffWithEffectTriggerCommand = void (*)(OnOffEffect *);
+
+    chip::EndpointId mEndpoint;
+    OffWithEffectTriggerCommand mOffWithEffectTrigger = nullptr;
+    uint8_t mEffectIdentifier;
+    uint8_t mEffectVariant;
+    bool mActive = false;
+
+    OnOffEffect(
+        chip::EndpointId endpoint, OffWithEffectTriggerCommand offWithEffectTrigger,
+        uint8_t effectIdentifier = static_cast<uint8_t>(EMBER_ZCL_ON_OFF_EFFECT_IDENTIFIER_DELAYED_ALL_OFF),
+        uint8_t effectVariant = static_cast<uint8_t>(EMBER_ZCL_ON_OFF_DELAYED_ALL_OFF_EFFECT_VARIANT_FADE_TO_OFF_IN_0P8_SECONDS));
```

Comment:
Why the casting instead of actually storing the right type?

Reply:
effectVariant type depends on the effect effectIdentifier so we don't know the type at compile time.
I can use effectIdentifier's real type but not the effectVariant.

Reply:
Worth a comment explaining that.

Reply:
effectIdentifier not fixed.

Reply:
fixed in #12813